Bug 1712072 - Update distribution docs to new structure (#1906)
mozilla/glean: docs/user/reference/metrics/memory_distribution.md

## Recording API

### `accumulate`

Accumulates the provided sample in the metric.

{{#include ../../../shared/tab_header.md}}

<div data-lang="Kotlin" class="tab">

```Kotlin
import org.mozilla.yourApplication.GleanMetrics.Memory

fun allocateMemory(nbytes: Int) {
    // ...
    Memory.heapAllocated.accumulate(nbytes / 1024)
}
```

</div>
<div data-lang="Java" class="tab">

```Java
import org.mozilla.yourApplication.GleanMetrics.Memory;

fun allocateMemory(nbytes: Int) {
    // ...
    Memory.INSTANCE.heapAllocated().accumulate(nbytes / 1024);
}
```

</div>
<div data-lang="Swift" class="tab">

```Swift
import Glean

func allocateMemory(nbytes: UInt64) {
    // ...
    Memory.heapAllocated.accumulate(nbytes / 1024)
}
```

</div>
<div data-lang="Python" class="tab">

```Python
from glean import load_metrics
metrics = load_metrics("metrics.yaml")

def allocate_memory(nbytes):
    # ...
    metrics.memory.heap_allocated.accumulate(nbytes / 1024)
```

</div>
<div data-lang="Rust" class="tab">

```rust
use glean_metrics::memory;

fn allocate_memory(bytes: u64) {
    // ...
    memory::heap_allocated.accumulate(bytes / 1024);
}
```

</div>

> <div data-lang="JavaScript" class="tab"></div>

Reviewer: ```suggestion
<div data-lang="JavaScript" class="tab" data-bug="1716952"></div>
```

<div data-lang="Firefox Desktop" class="tab">

**C++**

```c++
#include "mozilla/glean/GleanMetrics.h"

mozilla::glean::memory::heap_allocated.Accumulate(bytes / 1024);
```

**JavaScript**

```js
Glean.memory.heapAllocated.accumulate(bytes / 1024);
```

</div>

{{#include ../../../shared/tab_footer.md}}

#### Recorded errors

> * `invalid_value`: If recording a negative memory size.

Reviewer: ```suggestion
* [`invalid_value`](../../user/metrics/error-reporting.md): If recording a negative memory size.
```

Same below, or merge the two bullet points into one.

> * `invalid_value`: If recording a size larger than 1 TB.

Author: ```suggestion
* [`invalid_value`](../../user/metrics/error-reporting.md): If recording a size larger than 1 TB.
```

## Testing API

### `testGetValue`

> Gets the recorded value for a given counter metric.

Reviewer: ```suggestion
Gets the recorded value for a given memory distribution metric.
```

assertEquals(2L, snapshot.getCount());
```

</div>
<div data-lang="Swift" class="tab">

```Swift
@testable import Glean

// Get snapshot
let snapshot = try! Memory.heapAllocated.testGetValue()

// Does the sum have the expected value?
XCTAssertEqual(11, snapshot.sum)

// Usually you don't know the exact memory values,
// but how many should have been recorded.
XCTAssertEqual(2, snapshot.count)
```

</div>
<div data-lang="Python" class="tab">

```Python
from glean import load_metrics
metrics = load_metrics("metrics.yaml")

# Get snapshot.
snapshot = metrics.memory.heap_allocated.test_get_value()

# Does the sum have the expected value?
assert 11 == snapshot.sum

# Usually you don't know the exact memory values,
# but how many should have been recorded.
assert 2 == snapshot.count
```

</div>
<div data-lang="Rust" class="tab">

```rust
use glean::ErrorType;
use glean_metrics::memory;

// Get snapshot
let snapshot = memory::heap_allocated.test_get_value(None).unwrap();

// Does the sum have the expected value?
assert_eq!(11, snapshot.sum);

// Usually you don't know the exact timing values,
// but how many should have been recorded.
assert_eq!(2, snapshot.values.len());
```

</div>

> <div data-lang="JavaScript" class="tab"></div>

Reviewer: ```suggestion
<div data-lang="JavaScript" class="tab"  data-bug="1716952"></div>
```

<div data-lang="Firefox Desktop" class="tab">

**C++**

```c++
#include "mozilla/glean/GleanMetrics.h"

// Does it have an expected values?
const data = mozilla::glean::memory::heap_allocated.TestGetValue().value().unwrap()
ASSERT_EQ(11 * 1024, data.sum);
```

**JavaScript**

```js
const data = Glean.memory.heapAllocated.testGetValue();
Assert.equal(11 * 1024, data.sum);
```

</div>

{{#include ../../../shared/tab_footer.md}}

### `testHasValue`

> Whether or not **any** value was recorded for a given counter metric.

Reviewer: ```suggestion
Whether or not **any** value was recorded for a given memory distribution metric.
```

{{#include ../../../shared/tab_header.md}}

<div data-lang="Kotlin" class="tab">

```Kotlin
import org.mozilla.yourApplication.GleanMetrics.Pages

// Get snapshot
assertTrue(Memory.heapAllocated.testHasValue())
```

</div>

<div data-lang="Java" class="tab">

```Java
import org.mozilla.yourApplication.GleanMetrics.Pages;

// Was anything recorded?
assertTrue(Memory.INSTANCE.heapAllocated().testHasValue());
```

</div>
<div data-lang="Swift" class="tab">

```Swift
@testable import Glean

// Was anything recorded?
XCTAssert(Memory.heapAllocated.testHasValue())
```

</div>
<div data-lang="Python" class="tab">

```Python
from glean import load_metrics
metrics = load_metrics("metrics.yaml")

# Was anything recorded?
assert metrics.memory.head_allocated.test_has_value()
```

</div>
<div data-lang="Rust" class="tab"></div>
<div data-lang="JavaScript" class="tab"></div>
<div data-lang="Firefox Desktop" class="tab"></div>

{{#include ../../../shared/tab_footer.md}}

### `testGetNumRecordedErrors`

> Gets number of errors recorded for a given counter metric.

Reviewer: ```suggestion
Gets number of errors recorded for a given memory distribution metric.
```

<div data-lang="Kotlin" class="tab">

```Kotlin
import org.mozilla.yourApplication.GleanMetrics.Memory

// Did this record a negative value?
assertEquals(
    0,
    Memory.heapAllocated.testGetNumRecordedErrors(ErrorType.InvalidValue)
)
```

</div>
<div data-lang="Java" class="tab">

```Java
import org.mozilla.yourApplication.GleanMetrics.Memory;

// Assert that no errors were recorded.
assertEquals(
    0,
    Memory.INSTANCE.heapAllocated().testGetNumRecordedErrors(
        ErrorType.InvalidValue
    )
);
```

</div>
<div data-lang="Swift" class="tab">

```Swift
@testable import Glean

// Did this record a negative value?
XCTAssertEqual(0, Memory.heapAllocated.testGetNumRecordedErrors(.invalidValue))
```

</div>
<div data-lang="Python" class="tab">

```Python
from glean import load_metrics
metrics = load_metrics("metrics.yaml")

# Did this record a negative value?
assert 0 == metrics.memory.heap_allocated.test_get_num_recorded_errors(
    ErrorType.INVALID_VALUE
)
```

</div>
<div data-lang="Rust" class="tab">

```Rust
use glean::ErrorType;
use glean_metrics::pages;

assert_eq!(
    0,
    pages::page_load.test_get_num_recorded_errors(ErrorType::InvalidValue)
);
```

</div>

> <div data-lang="JavaScript" class="tab"></div>

Reviewer: ```suggestion
<div data-lang="JavaScript" class="tab"  data-bug="1716952"></div>
```

<div data-lang="Firefox Desktop" class="tab"></div>

{{#include ../../../shared/tab_footer.md}}

## Metric parameters

> Example custom distribution metric definition:

Reviewer: ```suggestion
Example memory distribution metric definition:
```
